Set up Avatars for Workspace Chats

marcochavezf opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR includes the avatars for workspace chats (for both admins and workspace members) and shows the workspace chats when the user has the `policyExpenseChat` beta. I also included fixes for some minor issues I found (like duplicate rows on the Search Page and the user name not included in the second line of the chat header).

Additionally, I created a Web-E PR to fix a few issues I found while I was implementing the workspace chats: https://github.com/Expensify/Web-Expensify/pull/33166

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/195471

### Tests / QA
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Pre-requisites:
  - Create an account which will be the admin
  - Create 2 or more accounts which will be the workspace members 
  - Add the created accounts to `policyExpenseChat` beta

Steps:
1. With the admin account create a Workspace and invite the workspace members.
2. Check the workspace chats were created for each workspace member in the Search Page and send them a message.
3. Login with the workspace members and verify there's only one workspace chat and check the message of the admin was received.

- [x] (Test) Verify that no errors appear in the JS console
- [ ] (QA) Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
User:
<img width="900" alt="Screen Shot 2022-02-23 at 18 41 44" src="https://user-images.githubusercontent.com/6829422/155435309-df1a152e-fa68-43a1-aba0-7f61d5ee5672.png">

Admin:
<img width="900" alt="Screen Shot 2022-02-23 at 18 43 15" src="https://user-images.githubusercontent.com/6829422/155435306-35de979e-b4c3-44aa-9ebc-2573c694041a.png">


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
User (with search value):
<img width="400" alt="Screen Shot 2022-02-23 at 18 49 21" src="https://user-images.githubusercontent.com/6829422/155435848-7c5fbcbd-853a-4b5d-8f5f-6c740e5fb848.png">

Admin (with search value):
<img width="400" alt="Screen Shot 2022-02-23 at 18 30 34" src="https://user-images.githubusercontent.com/6829422/155435384-56883e74-435b-452b-b6da-870fcb3ed77b.png">

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
<img width="900" alt="Screen Shot 2022-02-23 at 18 52 38" src="https://user-images.githubusercontent.com/6829422/155436158-bc28a27b-f53c-449f-84ab-4084bb13a412.png">


#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
<img width="400" alt="Screen Shot 2022-02-23 at 19 01 07" src="https://user-images.githubusercontent.com/6829422/155437156-56867e93-dca0-47e2-bc9f-f6451697d673.png">

<img width="400" alt="Screen Shot 2022-02-23 at 19 02 30" src="https://user-images.githubusercontent.com/6829422/155437157-84fc1883-68c9-4a5a-b5d1-61879f66e56b.png">


#### Android
<!-- Insert screenshots of your changes on the Android platform-->
<img width="400" alt="Screen Shot 2022-02-23 at 19 52 32" src="https://user-images.githubusercontent.com/6829422/155442233-745f8495-f520-4cc4-9484-0e777908298e.png">

<img width="400" alt="Screen Shot 2022-02-23 at 19 52 41" src="https://user-images.githubusercontent.com/6829422/155442237-1ee1252b-0947-4bb7-9fe2-6b64a5008c40.png">

marcochavezf committed:
check if the given report is a Workspace Chat

marcochavezf committed:
get the linked policy of the workspace chat

marcochavezf committed:
return avatar of the workspace for non-admin users

marcochavezf committed:
return participan and workspace avatar url for admins

marcochavezf committed:
use report.reportName for workspace chats

marcochavezf committed:
Merge branch 'main' into marco-avatarsWorkspaceChats

marcochavezf committed:
display workspace chat in search page

marcochavezf committed:
get reportName as option.text

marcochavezf committed:
expose option.isPolicyExpenseChat to use fullTitle

marcochavezf committed:
dislay alternateText for non-admins for workspace chats

marcochavezf committed:
change name param name in getSearchText

marcochavezf committed:
show workspace chat in search page for admins

marcochavezf committed:
get the correct report icons when user is admin

marcochavezf committed:
create styles for SubscriptAvatar

marcochavezf committed:
create SubscriptAvatar component

marcochavezf committed:
add showSubscript MultipleAvatars to render SubscriptAvatar

marcochavezf committed:
add showSubscript to option

marcochavezf committed:
update subscript avatar styles

TomatoToaster commented on src/libs/OptionsListUtils.js:
```diff
@@ -775,13 +780,29 @@ function getCurrencyListForSections(currencyOptions, searchValue) {
  * @param {Object} report
  * @param {Object} personalDetails
- * @returns {String}
+ * @returns {String[]}
```
Ah good catch this was always returning an array. But the typical format we use for this is

```suggestion
 * @returns {Array<String>}
```

marcochavezf replied:
Thanks @TomatoToaster :) I'm going to include it in the latest changes of today

TomatoToaster reviewed (changes requested):
Great start on this, excited for you to finish it. I just had one suggestion ahead of you finishing this.

marcochavezf committed:
show SubscriptAvatar in HeaderView

marcochavezf committed:
fix blank name in HeaderView for workspace chat

marcochavezf committed:
save ReportUtils function results in variables

marcochavezf committed:
fix duplicate workspace chates for workspace members

marcochavezf committed:
fix a few jsdoc errors

marcochavezf committed:
fix array of strings jsdoc

marcochavezf committed:
fix duplicate admin rooms in Search Page

marcochavezf committed:
display policy name as alternateText for admins

marcochavezf committed:
fix displayed name in alternateText for workspace chats

marcochavezf committed:
filter out workspace chat if user doesn't have permission

marcochavezf commented on src/libs/OptionsListUtils.js:
```diff
@@ -428,14 +458,15 @@ function getOptions(reports, personalDetails, activeReportID, {
-        if (logins.length <= 1) {
+        // Save the report in the map if this is a single participant (with the exception of PolicyExpenseChat and
+        // default rooms) so we can associate the reportID with the personal detail option later.
+        if (logins.length <= 1 && !isPolicyExpenseChat && !isDefaultRoom) {
```
I'm checking these flags to prevent a bug where the policy and policyExpense (workspace) chats were included twice in the Search Page

marcaaron replied:
What was the bug and can we maybe say a bit more about it? We added to the comment...

>(with the exception of PolicyExpenseChat and default rooms)

It is not easy to see why we are making this exception. If I saw it I would just assume whoever wrote this had a good reason and stay away from it 😂

marcochavezf replied:
I was having duplicated workspace chats, but that's now fixed with the participants' change I made [here](https://github.com/Expensify/App/pull/7852/files/6ad97fb3d52ef67c3d11b4506f9d802d83e2161e#r815350015), also I thought the same was happening with default rooms but now with more context, I realized this is not the case. I'm going to revert this condition and the comment. Good catch and thanks for pointing that!

marcochavezf commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +204,41 @@ function hasReportDraftComment(report) {
+ * @param {Object} report
+ * @returns {Array<String>}
+ */
+function getParticipants(report) {
+    const participants = [...lodashGet(report, ['participants'], [])];
+    if (ReportUtils.isPolicyExpenseChat(report)) {
+        if (!participants.includes(report.ownerEmail)) {
+            participants.push(report.ownerEmail);
+        }
+
+        // Include admin as participant
+        if (!report.isOwnPolicyExpenseChat) {
+            participants.push(currentUserLogin);
+        }
+    }
+    return participants;
+}
```
I created this function to include the admin and/or the workspace member as participants in order to include the name on the second line of the chat header (i.e. "User1: test message").

marcaaron replied:
I think the admin should be a participant. Any ideas on why they not showing up in the participants array?

marcaaron replied:
Report owner makes sense as they do not have a `sharedReports` record for their own policy expense chat.

marcochavezf replied:
Ohh that's right, I realized it was excluded here: https://github.com/Expensify/App/blob/36c6f8a2ccb4a846daff664036571c2cdf1fc164/src/libs/actions/Report.js#L126

I fixed it and removed this condition in `getParticipants`

TomatoToaster reviewed (changes requested):
Actually there's one more thing I'm realizing we need to add. We need to show a better version of the welcometext we see here:
<img width="441" alt="image" src="https://user-images.githubusercontent.com/19364431/155595501-0b87a24e-b70a-423f-a2c8-0d9e27abfbcf.png">


I'll detail where we need to change this.

TomatoToaster commented:
So here: https://github.com/Expensify/App/blob/ce055a578be36ef5e4a28badb5187a17b2b63484/src/pages/home/report/ReportActionItemCreated.js#L34

We'll want to add a `showSubscript` field and have it be equal to `isPolicyExpenseChat && !isPolicyExpenseChat` like you do in other places

Then right after here: https://github.com/Expensify/App/blob/1d10942b60b5e8e6d8db3ac63b38fe8f914acc39/src/components/RoomHeaderAvatars.js#L36-L37
if `props.showSubscript` then we'll show a SubscriptAvatar BUT lets add a size field and give a value of `large`.


Then we'll check if the size is large here in SubscriptAvatar: https://github.com/Expensify/App/blob/39cc2868f171528644c616fe65233b06b97b851b/src/components/SubscriptAvatar.js#L20

And if it's large then we'll use the large avatar style instead of `styles.avatarNormal`. We might also have to tweak some special styling for the subscript avatar in the large case but I'm not sure yet.

Could you add this to this PR too?

TomatoToaster commented on src/components/SubscriptAvatar.js:
```diff
@@ -0,0 +1,40 @@
+
+    /** Tooltip for the Avatar */
+    avatarTooltips: PropTypes.arrayOf(PropTypes.string).isRequired,
+};
```
We will want to show the `Expensicons.workspace` icon here if  `avatarImageURLs[1] ===''` which will be the case when the users' workspace does not have an avatar set. Without this we'll see this:
<img width="58" alt="image" src="https://user-images.githubusercontent.com/19364431/155604216-77c37b2b-1c9d-42c2-b683-02a505de4ad4.png">


So here let's add a 'defaultSubscriptIcon' prop here that will be of type `PropTypes.func.building.isRequired`. Then we'll conditionally render the icon  if `avatarImageURLs[1] ===''` using `<Icon src ={defaultSubscriptIcon} ... />`. To get the the subscript style you'll have to wrap this in a View and apply the style on that. You'll also have to set the icon's width and height manually.

CC: @marcaaron, @roryabraham if you think this is solid or there is a better way to do this.

TomatoToaster replied:
Then in places where we call <SubscriptAvatar> for the policyExpenseChats we'll set `Expensicons.workspace` as the `defaultSubscriptIcon` prop.

marcaaron replied:
What is the significance of `avatarImageURLs[1]`? The user at the 1st index will be an empty string seems a little magical and subject to change unintentionally. Can we maybe just pass a prop that says whether to use this icon or not? Do we have a better way to tell this is a workspace chat and should use this icon?

roryabraham replied:
> So here let's add a 'defaultSubscriptIcon' prop here ... [then] conditionally render the icon if avatarImageURLs[1] ==='' using <Icon src ={defaultSubscriptIcon} ... />

I have two suggestions here. First, this `SubscriptAvatar` component can only really have two avatars right? So instead of using an `avatarImageURLs` prop that's an array, where the "main" avatar is at index 0 and the "subscript" avatar is at index 1, why not just have separate props for the `mainAvatar` and the `secondaryAvatar`?

Second, I think we could clean this up further by refactoring the `Avatar` component to make it a bit more reusable. Right now it takes an image URL, and also `isChatRoom` and `isArchivedRoom`, which are used to display static icon assets for chat rooms (which don't have image avatars yet AFAIK). Instead, `Avatar` could more generically take:

1. A source, which can be either an image URL or a static icon
1. A fallback icon, which can only be static icon (available offline).
    
So a simplified version might look something like this:

```js
const propTypes = {
    /** Source for the avatar. Can be the URL of an image or an icon. */
    source: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
    
    /** Icon to render if the source image could not be loaded. */
    fallbackIcon: PropTypes.func,
    
    /** Information about the network */
    network: PropTypes.shape({
        /** Is the network currently offline or not */
        isOffline: PropTypes.bool,
    }),
};

const defaultProps = {
    fallbackIcon: () => {},
    network: {},
};

const Avatar = (props) => {
    if (!props.source || props.network.isOffline) {
        return props.fallbackIcon
            ? <Icon src={props.fallbackIcon} />
            : null;
    }

    return (
        <View style={props.containerStyles}>
            {
                _.isFunction(props.source)
                    ? <Icon src={props.source} />
                    : <Image source={{uri: props.source}} />
            }
        </View>
    );
};
```

I think that could simplify things here and elsewhere.

marcochavezf committed:
add showSubscript to RoomHeaderAvatars

marcochavezf committed:
show workspace icon for subscript avatar

marcochavezf committed:
add defaultSubscriptIcon to SubscriptAvatar

marcochavezf committed:
set Expensicons.Workspace in defaultSubscriptIcon

marcochavezf committed:
show SubscriptAvatar component in RoomHeaderAvatars

marcochavezf committed:
create emptyAvatarLarge and add var avatarSizeLarge

marcochavezf committed:
add size to SubscriptAvatar and use Avatar instead of Image

marcaaron commented on src/components/MultipleAvatars.js:
```diff
@@ -63,6 +68,15 @@ const MultipleAvatars = (props) => {
+                avatarTooltips={props.avatarTooltips}
+            />
+        );
+    }
```
I had a question about why we should use `MultipleAvatars`  component just to return a different component when a boolean prop is present? Does it make sense to just use this component directly?

marcochavezf replied:
Oh not having dependency with `MultipleAvatars` makes sense, I removed it from `MultipleAvatars` and call it directly.

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -274,13 +295,15 @@ function createOption(personalDetailList, report, {
         iouReportAmount: lodashGet(iouReport, 'total', 0),
         isChatRoom,
+        isPolicyExpenseChat,
+        showSubscript: isPolicyExpenseChat && !report.isOwnPolicyExpenseChat,
```
```suggestion
        shouldShowSubscript: isPolicyExpenseChat && !report.isOwnPolicyExpenseChat,
```

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -311,7 +334,7 @@ function isSearchStringMatch(searchValue, searchText, participantNames = new Set
  * Returns the given userDetails is currentUser or not.
  * @param {Object} userDetails
- * @returns {Bool}
+ * @returns {Boolean}
```
❤️

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -775,13 +806,29 @@ function getCurrencyListForSections(currencyOptions, searchValue) {
 function getReportIcons(report, personalDetails) {
     // Default rooms have a specific avatar so we can return any non-empty array
     if (ReportUtils.isChatRoom(report)) {
         return [''];
```
I know this wasn't added here, but anyone know what is up with the returning an array with an empty string? What is that supposed to mean and why should it be "non-empty"?

TomatoToaster replied:
I believe we return an empty string because we're using an Icon component instead of using a url for an Image when we're rendering the Avatars for rooms. This could be null, or empty or I guess we could pass the Icon component itself here, but the way we built it just has [RoomAvatar.js](https://github.com/Expensify/App/blob/33e5d99b95e8e7f767c42f5dccbb76879a565387/src/components/RoomAvatar.js#L20) just ignore whatevers stored here anyway. I think having this be empty string satisfies some propType checks.

marcaaron replied:
Ok thanks for the explanation. Maybe we can return an empty array or allow the icons to be `undefined` and not have to say anything about how the report icons will be used or why the array must be non empty. But don't think we have to block this PR on it though - just something I was curious about.

roryabraham replied:
@marcochavezf Did our changes to `Avatar` change this?

marcochavezf replied:
Ah no, this `getReportIcons` is called before we save reports into Onyx and we can't set the corresponding Icon components here, as I suggested in this [comment](https://github.com/Expensify/App/pull/7852#discussion_r817152760) I think (in order to move/merge the logic of `getAvatarSources` to `getReportIcons`) we'll need to implement a different approach or maybe remove `getReportIcons` from `PersonalDetails`: 

https://github.com/Expensify/App/blob/main/src/libs/actions/PersonalDetails.js#L192-L211 

But I can address it in another ticket if it's ok because my gut tells me this can break something and we'll require more time to address it :D

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -775,13 +806,29 @@ function getCurrencyListForSections(currencyOptions, searchValue) {
+            `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`, 'avatarURL',
+        ]);
+
+        // If the user is not an admin for this workspace chat, return avatar of the workspace
```
This comment would be clearer if it doesn't mention anything about admins. The user can be an admin AND this could be their "own policy expense chat".

marcochavezf committed:
add secondAvatarLarge and firstAvatar styles

marcochavezf committed:
apply different styles if size is large

marcochavezf committed:
set large size in RoomHeaderAvatars

marcochavezf commented:
I fixed the avatars in `RoomHeaderAvatars` and when there's no image in `SubscriptAvatar` (use workspace icon by default):

<img width="900" alt="Screen Shot 2022-02-24 at 18 40 33" src="https://user-images.githubusercontent.com/6829422/155632743-7e251921-2d5a-4a10-9999-c582d1c89dc4.png">

<img width="900" alt="Screen Shot 2022-02-24 at 18 45 41" src="https://user-images.githubusercontent.com/6829422/155632748-3061c59c-390a-4d0c-ae4c-3d33654ff6b3.png">

But I still need to fix the scenario when there's only one avatar and there's no image:

<img width="350" alt="Screen Shot 2022-02-24 at 18 52 38" src="https://user-images.githubusercontent.com/6829422/155632880-646fac65-31fd-443d-827f-0fa7c8857edc.png">

I think the best solution for that is to create a new component `PolicyExpenseChat` (similar to `RoomAvatar`) and add it to `Avatar` here:

https://github.com/Expensify/App/blob/d586981aa5aeb607c3f2cb7dee5d3e251fa7421c/src/components/Avatar.js#L54-L56

Also tomorrow I address the remaining comments.

TomatoToaster commented on src/components/RoomHeaderAvatars.js:
```diff
@@ -6,25 +6,42 @@ import styles from '../styles/styles';
+                avatarImageURLs={props.avatarImageURLs}
+                avatarTooltips={[]}
+                defaultSubscriptIcon={() => Expensicons.Workspace}
+                size="large"
```
Yes agreed let's use this https://github.com/Expensify/App/blob/737cc12e69d6537372baa1c7dee08aaa353dcf2f/src/CONST/index.js#L545-L548

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +204,41 @@ function hasReportDraftComment(report) {
+ * @returns {Array<String>}
+ */
+function getParticipants(report) {
+    const participants = [...lodashGet(report, ['participants'], [])];
```
Let's test it? 90% sure can be 100% wrong :trollface:

marcochavezf replied:
Ah yeah, I added on purpose because otherwise we mutate `report.participants` and we end up with tons of repeated participants when we add the admin and/or report owner a few lines below.

marcochavezf committed:
move chat room icons to Expensicons

marcochavezf committed:
use Icon instead of RoomAvatar in Avatar

marcochavezf committed:
delete RoomAvatar (logic moved to Avatar)

marcochavezf committed:
display archived room in RoomHeaderAvatar

marcochavezf committed:
fix border for large avatar in RoomHeaderAvatar

marcochavezf committed:
create LargeDualAvatars to remove complexity in SubscripAvatar

marcochavezf committed:
remove LargeDualAvatars from RoomHeaderAvatars and call it directly

marcochavezf committed:
remove size from SubscriptAvatar

marcochavezf committed:
add new avatar sizes

marcochavezf committed:
add size prop to Avatar to re-use it as Icon

marcochavezf committed:
use size property to remove Icon from LargeDualAvatars

marcochavezf committed:
display workspace icon for empty workspace chats

marcochavezf committed:
display workspace icon in OptionRow and RoomHeaderAvatar

marcochavezf commented:
I fixed the empty workspace chat icon and addressed some of the changes in the comments. 

<img width="900" alt="Screen Shot 2022-02-25 at 17 35 14" src="https://user-images.githubusercontent.com/6829422/155819043-7c2d9c80-4d8a-4db2-ad95-aced6cf65165.png">

Tomorrow I will continue to respond the comments and address a few outstanding changes.

marcochavezf committed:
Merge branch 'main' into marco-avatarsWorkspaceChats

marcochavezf committed:
add functions getAvatarSource

marcochavezf committed:
use OptionsListUtils.getAvatarSource

marcochavezf committed:
remove logic to check avatar source from Avatar

marcochavezf committed:
get avatar sources (url or icon) for OptionRow

marcochavezf committed:
rename avatarImageURLs to avatarIcons and change props

marcochavezf committed:
remove unused isChatRoom, isArchivedRoom and isPolicyExpenseChat

marcochavezf committed:
change params to SubscriptAvatar

marcochavezf committed:
use _.contains() instead of .includes()

marcochavezf committed:
include participants in getParticipantEmailsFromReport

marcochavezf committed:
change showSubscript flag to shouldShowSubscript

marcochavezf committed:
revert condition for reportMapForLogins (fixed with report.participants fix)

marcochavezf committed:
give priority to isArchivedRoom in getAvatarSource

marcochavezf committed:
update comment about returning workspace avatar

marcochavezf committed:
use props.source directly

marcochavezf committed:
move large dual avatars to RoomHeaderAvatars and remove LargeDualAvatars

marcochavezf committed:
fix jsdoc for icons

marcochavezf commented:
@marcaaron @roryabraham @TomatoToaster Hi guys! I addressed some of the requested changes and the PR is ready to be reviewed again.

I followed @roryabraham's advice of using `source: PropTypes.oneOfType([PropTypes.string, PropTypes.func])` in `Avatar` to make it more generic and looks good! 

I additionally created a couple of functions (`getAvatarSourceFromReport` and `getAvatarSource`) in `OptionsListUtils` to get the icons according to the `chatType` (ChatRoom, ArchivedRoom and PolicyExpenseChat). These functions return an array of URLs and/or icons ready to be consumed by the components that use `Avatar` and in this way we avoid passing the params `isChatRoom`, `isPolicyExpenseChat`, etc. through the intermediate components that use `Avatar`.

With these changes, the `Avatar` is more re-usable and proof of that is the removal of the component `RoomAvatar`, which was replaced by 2 conditions in `getAvatarSource`:
https://github.com/Expensify/App/blob/40325c727ed7d0e3dfa4f3edf6eaa2a0d673e8e7/src/libs/OptionsListUtils.js#L234-L238

I also renamed `avatarImageURLs` to `avatarIcons` to make it consistent and changed its propTypes to `PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.func]))` in some places.
 
I didn't add the `isNetworkOffline` and `fallbackIcon` in `Avatar` in this PR because I think that will require a little more time to test it and can keep blocking the next N7 issues, but I can add it as a follow-up PR. Also, I think some of the logic of `RoomHeaderAvatars` and `MultipleAvatars` can be refactored/merged, but I can also do it in a follow-up PR.

marcaaron commented on src/components/Avatar.js:
```diff
@@ -17,43 +21,45 @@ const propTypes = {
+            [CONST.AVATAR_SIZE.SMALL]: variables.avatarSizeSmall,
+            [CONST.AVATAR_SIZE.LARGE]: variables.avatarSizeLarge,
+        };
+        const iconSize = AVATAR_SIZES[this.props.size];
```
NAB, but this logic feels like it could be in [`StyleUtils`](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) as a `getAvatarSize()` or something

marcaaron commented on src/components/RoomHeaderAvatars.js:
```diff
@@ -6,46 +6,69 @@ import styles from '../styles/styles';
+        );
+    }
+
+    // avatarIconsToDisplay
```
Delete?

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,76 @@ function hasReportDraftComment(report) {
+ */
+function getParticipants(report) {
+    const participants = [...lodashGet(report, ['participants'], [])];
+    if (ReportUtils.isPolicyExpenseChat(report)) {
+        if (!_.contains(participants, report.ownerEmail)) {
+            participants.push(report.ownerEmail);
+        }
+    }
+    return participants;
```
```suggestion
    if (ReportUtils.isPolicyExpenseChat(report) && !_.contains(participants, report.ownerEmail)) {
        participants.push(report.ownerEmail);
    }
    return participants;
```

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,76 @@ function hasReportDraftComment(report) {
+ * @param {Boolean} [options.isChatRoom]
+ * @param {Boolean} [options.isArchivedRoom]
+ * @param {Boolean} [options.isPolicyExpenseChat]
+ * @returns {String | function():any}
```
How about just `returns {*}`? I think most places we have standardized on using the capitalized docs. Not sure I have seen `function():any` used before or understand what it means 😬 

Caveat...

1. We have style guide recommendations here for JSDocs https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs
2. They could probably be better so please feel free to propose changes
3. Probably would be best to follow the style guide until we have done Step 2 to avoid blocking

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,76 @@ function hasReportDraftComment(report) {
+ * @param {Boolean} [options.isPolicyExpenseChat]
+ * @returns {String | function():any}
+ */
+function getAvatarSource(source, {isChatRoom, isArchivedRoom, isPolicyExpenseChat}) {
```
NAB, I think this would be better if we pass a type as a string instead of an object with the different boolean props (they look like they are mutually exclusive so not sure why we would pass an object of one or more of them).

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,76 @@ function hasReportDraftComment(report) {
+ * @returns {String | function():any}
+ */
+function getAvatarSource(source, {isChatRoom, isArchivedRoom, isPolicyExpenseChat}) {
+    if (!source) {
```
Why do we always want to return the `source` if we have one? Is there some case where we are specifying the kind of avatar that we want to use and also providing a `source`?

marcaaron replied:
Asked another way, can we just return the source when we haven't specified a generic type?

marcochavezf replied:
I merged these functions into one function that only receives the `report` object, I think it's now cleaner

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,76 @@ function hasReportDraftComment(report) {
+ * Get the Avatar urls or fallback to the default icons according to the chat type
+ *
+ * @param {Object} report
+ * @returns {Array<String | function():any>}
```
Doc needs fixing

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,76 @@ function hasReportDraftComment(report) {
+    return _.map(report.icons, source => getAvatarSource(source, {
+        isChatRoom: ReportUtils.isChatRoom(report),
+        isArchivedRoom: ReportUtils.isArchivedRoom(report),
+        isPolicyExpenseChat: ReportUtils.isPolicyExpenseChat(report),
```
Should we pass the report and have this logic inside `getAvatarSource()`?

timszot commented:
Removing myself as a reviewer here since there are three others reviewing.

marcochavezf committed:
move avatar size logic to the new funciton getAvatarSize

marcochavezf committed:
delete redundant comment

marcochavezf committed:
move condition logic to one line

marcochavezf committed:
use {*} to follow jsdoc style guides

marcochavezf committed:
merge logic of getAvatarSources

marcochavezf committed:
replace getAvatarSourceFromReport with getAvatarSources

marcochavezf committed:
return Profile icon instead of personalDetail.avatar

marcochavezf commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,58 @@ function hasReportDraftComment(report) {
+        if (ReportUtils.isPolicyExpenseChat(report)) {
+            return Expensicons.Workspace;
+        }
+        return Expensicons.Profile;
```
I added `Expensicons.Profile` to remove `[personalDetail.avatar]` from `createOption` function: 

https://github.com/Expensify/App/blob/00805b2b495fac25f53a42231dd3e5ae78138a2a/src/libs/OptionsListUtils.js#L259

I realized that `personalDetail.avatar` is not the correct solution when the icon is still not present (which happens when the app is bootstrapping and loading the reports for the first time) because `personalDetail` is the first option of the array `personalDetails` and the first option not always correspond to the real report avatar (in some cases it shows for a second the avatar of the user logged in and then the real avatar).

marcochavezf committed:
fix styles of large workspace room avatars

marcochavezf committed:
call SubscriptAvatar directly

marcochavezf commented:
I updated the PR with the requested changes, also I retested the app on iOS and Android, and looks good.

marcochavezf committed:
set avatarTooltips as optional prop

marcochavezf committed:
fix tooltips in SubscriptAvatar

marcochavezf commented:
Update: I noticed I missed Rory's comment related to the tooltips and I realized the tooltips were not shown on the workspace avatars. I pushed some changes to address that and I re-tested again on all platforms. The PR is ready to be reviewed again 🙇🏽‍♂️ cc @marcaaron @roryabraham @TomatoToaster

roryabraham commented on src/components/SubscriptAvatar.js:
```diff
@@ -0,0 +1,55 @@
+
+const SubscriptAvatar = props => (
+    <View style={styles.emptyAvatar}>
+        <View>
```
Is this `View` necessary?

marcochavezf replied:
Oh good catch, I'm going to remove it

roryabraham commented on src/components/RoomHeaderAvatars.js:
```diff
@@ -6,46 +6,68 @@ import styles from '../styles/styles';
+                        fill={themeColors.iconSuccessFill}
+                    />
+                </View>
+                <View style={[styles.secondAvatarHovered, styles.rightSideLargeAvatar]}>
```
I've looked into it, and it seems like `secondAvatarHovered` is a poor name for this style:

```js
secondAvatarHovered: {
    backgroundColor: themeColors.sidebarHover,
    borderColor: themeColors.sidebarHover,
}
```

It's used in several locations that don't have much to do with hovering. What this style does is provide the white background around the secondary avatars to create a visual space between avatars.

<img width="118" alt="image" src="https://user-images.githubusercontent.com/47436092/156233705-12d50d00-e996-460d-9aee-7405181e1490.png">

So I would rename it to something else that's more accurately representative of the style. One option would be to just rename it to `secondAvatarBackground`, but another option would be to rename `defaultModalContainer` to just `defaultComponentBG` and reuse that in place of `secondAvatarHovered`.

roryabraham replied:
Another option would be to use `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`

marcochavezf replied:
Oh makes sense, I'm going to replace `secondAvatarHovered` with `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`

roryabraham commented on src/components/SubscriptAvatar.js:
```diff
@@ -0,0 +1,55 @@
+);
+
+SubscriptAvatar.propTypes = propTypes;
+SubscriptAvatar.defaultProps = defaultProps;
```
Looks like you forgot `displayName` on this component

roryabraham commented on src/pages/ReportDetailsPage.js:
```diff
@@ -132,11 +132,9 @@ class ReportDetailsPage extends Component {
                                 containerStyles={[styles.singleAvatarLarge, styles.mb4]}
                                 imageStyles={[styles.singleAvatarLarge]}
-                                source={this.props.report.icons[0]}
+                                source={OptionsListUtils.getAvatarSources(this.props.report)}
```
`OptionsListUtils.getAvatarSources` returns an array, right? But `Avatar`'s `source` prop should be a string or function. Something seems wrong here ... should it be like this?

```suggestion
                                source={_.first(OptionsListUtils.getAvatarSources(this.props.report))}
```

marcochavezf replied:
Oh good catch, I think I missed this when I merged `getAvatarSources` functions. Not sure why I didn't see an error when I tested

roryabraham commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,58 @@ function hasReportDraftComment(report) {
+ * @param {Object} report
+ * @returns {Array<*>}
+ */
+function getAvatarSources(report) {
```
It seems like this might not be needed. Could we put all the report-type-specific `ReportUtils.isXYZ` checks in `OptionsListUtils.getReportIcons`? 

And if that's the case, this function would end up looking like:

```js
function getAvatarSources(report) {
    return _.map(lodashGet(report, 'icons', []), source => source || Expensicons.Profile);
}
```

marcochavezf replied:
Ah I tried to implement the avatar sources logic in `getReportIcons` before but it's not possible because `getReportIcons` is called to update the reports to then save them into Onyx here:

https://github.com/Expensify/App/blob/daafe503149ff80c221a54d8df2b24082dcf9d42/src/libs/actions/PersonalDetails.js#L192-L211

And this approach was throwing errors because the icons (functional components) can't be saved in the storage.

I think we'll need a more holistic / different way to solve this (maybe we can return an object in `getReportIcons` to represent the icon and map that object to the corresponding Icon component just when the `Avatar` is rendered?) because also we're using an empty string to workaround this case in `getReportIcons`: https://github.com/Expensify/App/pull/7852/files#r814365077

I can create a ticket for that to discuss it deeper and address it later if it's ok :)

roryabraham replied:
Yep, thanks for the explanation. Sounds like  better idea to address in a follow-up. Can you please create the issue since you have more context and cc me?

roryabraham replied:
Also, if you don't think it's worth it we can just :donothing: and move on. I'll leave that up to you.

marcochavezf replied:
This is the ticket https://github.com/Expensify/Expensify/issues/199856

roryabraham commented on src/pages/home/sidebar/OptionRow.js:
```diff
@@ -163,22 +164,29 @@ const OptionRow = (props) => {
-                                        avatarTooltips={avatarTooltips}
-                                    />
+                                    props.option.shouldShowSubscript ? (
+                                        <SubscriptAvatar
```
Have you tested this in #focus mode? Just looking at how `MultipleAvatars` is used in this component, and it looks like there was a need to adjust the size of the avatar if `props.mode === 'compact'`

marcochavezf replied:
Ohh I missed that, checking that. Also, I'm kind of confused about the #focus mode, this only changes the size, right? because noticed in the production app that chats disappear from LHN and only the Chronos chat is shown

roryabraham replied:
> Also, I'm kind of confused about the #focus mode, this only changes the size, right? because noticed in the production app that chats disappear from LHN and only the Chronos chat is shown

Yeah, there's more that changes with #focus mode. I think it only shows chats for which you have unread messages

marcochavezf committed:
Update src/components/Avatar.js

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

marcochavezf committed:
remove unnecessary View in SubscriptAvatar

marcochavezf committed:
replace secondAvatarHovered with StyleUtils.getBackgroundAndBorderStyle

marcochavezf committed:
add missing displayName

marcochavezf committed:
get the first element of avatar sources in ReportDetailsPage

marcochavezf committed:
change SubscriptAvatar size in #focus mode

marcochavezf committed:
use Avatar size const instead in MultipleAvatars

marcochavezf commented:
Update: I addressed the requested changes and the PR is ready to be reviewed again. One minor issue is the styling of workspace avatars in #focus mode, but I think this is not a blocker for the next N7 issues and I can address it in a follow-up PR (along with the `getReportIcons` issue discussed in the last comments).

<img width="800" alt="Screen Shot 2022-03-01 at 18 33 52" src="https://user-images.githubusercontent.com/6829422/156272384-91cb5000-f1db-4315-b8a5-dd57bae5c059.png">

roryabraham reviewed (dismissed):
Personally I would prefer to see the style issues with the `SubscriptAvatar` component fixed in this PR, but I won't block you if you want to address that in a follow-up.

TomatoToaster reviewed (dismissed):
I think it's fine if you want to work on focus mode styles in a follow up PR. However I've already gotten started on the issue that this PR would be blocking, and I've branched off this PR already, so if you want to finish in this PR that's fine too.

If you do end up doing this another PR, just make sure it doesn't close the original issue when this PR is merged!

marcochavezf commented:
> I think it's fine if you want to work on focus mode styles in a follow up PR. However I've already gotten started on the issue that this PR would be blocking, and I've branched off this PR already, so if you want to finish in this PR that's fine too.
> 
> If you do end up doing this another PR, just make sure it doesn't close the original issue when this PR is merged!

Ohh cool, I was a little worried about blocking the next issue. And for the #focus mode I noticed I will need to do more changes (create new styles, a new size for the subscript and do a little of clean-up) and can extend the time to merge this PR. But it's great to know that this not a blocker atm, so I'm going to continue with the #focus mode changes in this same PR today 😀

marcochavezf committed:
create getAvatarStyle function and add SMALL_SUBSCRIPT size

marcochavezf committed:
use StyleUtils.getAvatarStyle in Avatar

marcochavezf committed:
create OPTION_MODE and replace string literals

marcochavezf committed:
add mode to SubscriptAvatar and secondAvatarSubscriptCompact style

marcochavezf committed:
set emptyAvatarSmall and update style for subscript avatar

marcochavezf commented:
Update: I pushed the changes to handle the #focus mode styles in SubscriptAvatar:

<img width="800" alt="Screen Shot 2022-03-02 at 14 08 15" src="https://user-images.githubusercontent.com/6829422/156450398-0d5daaa5-45ec-49f6-bf9d-3d493acd1ec0.png">

<img width="800" alt="Screen Shot 2022-03-02 at 14 08 38" src="https://user-images.githubusercontent.com/6829422/156450403-967f3090-13d2-4655-83e8-e92015236e2e.png">

The PR is ready to be reviewed again 😁

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -204,20 +205,58 @@ function hasReportDraftComment(report) {
+ * @returns {Array<*>}
+ */
+function getAvatarSources(report) {
+    return _.map(lodashGet(report, 'icons', ['']), (source) => {
```
If can, please add some kind of comment here about why we are defaulting to an array with an empty string.

marcaaron reviewed (approved):
This LGTM just one had more comment to leave. 

Very nice work overall with this and I appreciate the clean up that is happening here. I would maybe have liked to see a PR with fewer changes (though not sure how possible it would have been in this case). Did find it a bit hard to focus on all the changes at once.

marcaaron commented:
This is some good advice [here](https://github.com/Expensify/Web-Expensify/blob/main/_devportal/docs/pullRequest.md#make-it-easy-for-reviewers) (at least I realized how important it is as time goes on).

TomatoToaster commented on src/pages/ReportDetailsPage.js:
```diff
@@ -132,11 +132,9 @@ class ReportDetailsPage extends Component {
                                 containerStyles={[styles.singleAvatarLarge, styles.mb4]}
                                 imageStyles={[styles.singleAvatarLarge]}
-                                source={this.props.report.icons[0]}
+                                source={_.first(OptionsListUtils.getAvatarSources(this.props.report))}
```
I wanna note that this breaks the icons here, but I've already fixed that in my PR with this commit (I used your updates to RoomHeaderAvatars too!): https://github.com/Expensify/App/pull/7957/commits/50300783f38c2b6d5a3397b4925ec070751e1305

It's not really relevant in this PR, so no need to add it here, I just noticed it when I added reportDetailsView for the policyExpenseChats (which is something I have to do for my PR). The only reason I bring it up is that if anyone notices this is off, it's already fixed!

TomatoToaster reviewed (approved):
You can add the comment that Marc mentioned but otherwise this LGTM.

roryabraham commented on src/components/SubscriptAvatar.js:
```diff
@@ -0,0 +1,61 @@
+    secondaryTooltip: PropTypes.string,
+
+    /** Set the size of avatars */
+    mode: PropTypes.oneOf(_.values(CONST.OPTION_MODE)),
```
NAB but in your follow-up I would recommend changing this prop to `size` to match the other avatar components

marcochavezf replied:
Sure thing, also I will add a Storybook stories of SubscriptAvatar

marcochavezf replied:
I addressed this change here https://github.com/Expensify/App/pull/8144

botify commented:
@roryabraham looks like this was merged without passing tests. Please add a note explaining why this was done and remove the `Emergency` label if this is not an emergency.

roryabraham commented:
I'm like 90% sure tests were already passing here. Unless I'm going crazy I've seen this `Emergency` label get mis-applied a few times recently

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

thesahindia commented:
Looks like we have got a regression
<img width="369" alt="Screenshot 2022-03-04 at 1 52 26 PM" src="https://user-images.githubusercontent.com/83179295/156726406-8d843cc0-f549-4da9-9db0-4ff30d412df4.png">
cc: @roryabraham @marcochavezf

TomatoToaster commented:
@thesahindia I've got a fix for this, let me pull it out of my PR so it can be used here.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1965977435) to staging by @roryabraham in version: 1.1.42-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

TomatoToaster commented:
Verified this worked on web:
user view:
<img width="1305" alt="image" src="https://user-images.githubusercontent.com/19364431/158441049-85fa732c-3cef-4ba6-9df6-7dc8bb32837e.png">

From admin view:
<img width="1302" alt="image" src="https://user-images.githubusercontent.com/19364431/158441126-fc5ab600-7bbf-4c75-931c-f899db886fe8.png">

Admin's own chat:
<img width="1296" alt="image" src="https://user-images.githubusercontent.com/19364431/158441196-609d0f4c-0ef6-4c60-82d1-6c46cac89005.png">

iOS images in next comment.

TomatoToaster commented:
iOS images:
admin chat view:
![image](https://user-images.githubusercontent.com/19364431/158441417-9c0594c4-166c-412e-94e0-9b203ece7424.png)

admin viewing user's chat:
![image](https://user-images.githubusercontent.com/19364431/158441514-fbfe53f1-ec23-45ed-9013-4e9eea1b1b41.png)

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1989053438) to production by @chiragsalian in version: 1.1.42-6 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
